[fx graph cache] Support freezing with FX graph caching

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136505

Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation: If freezing has not been applied, then keep the existing behavior (pickle the metadata and values). If freezing has been applied, then pickle the values if the constant will be inlined; otherwise, consider only the metadata.
2) If freezing has been applied, modify what we store in the cache: Instead of storing the constant attributes in the cache entry, store the _names_ of the constants, and then grab those constants from the GraphModule when we need attache the attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along.

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

masnesral committed:
[inductor] Support freezing with FX graph caching

Summary:

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary:

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

[ghstack-poisoned]

masnesral committed:
fix tests on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

[ghstack-poisoned]

eellison commented on test/inductor/test_codecache.py:
```diff
@@ -334,18 +338,18 @@ def fn(x):
     @parametrize("device", (GPU_TYPE, "cpu"))
     def test_constant_handling(self, device):
```
nice. could also parameterize by size and have one < 8 and one >

masnesral replied:
Yeah, can do!

eellison commented on test/inductor/test_codecache.py:
```diff
@@ -573,13 +577,48 @@ def f(x, val):
         Test the hashing of tensor constants.
         """
-        data = FxGraphCachePickler.dumps(torch.tensor(list(range(9))))
+        data = FxGraphCachePickler.dumps(torch.tensor(list(range(8))))
```
worth it to parameterize here too ?

eellison reviewed (approved):
looks great !! I think you set off benchmark run, but let's wait for the dashboard to show green before merging, and make sure to check a run with freezing & inference enabled.

cc @jamesjwu, we do still need to skip aot autograd caching if freezing is enabled. don't know if there is a test already for this but if not could we add ?

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -1256,10 +1253,9 @@ def _check_can_cache(gm: torch.fx.GraphModule) -> None:
         """
-        # Freezing can embed constants that wouldn't be static across runs.
-        if config.freezing or config.aot_inductor.use_runtime_constant_folding:
+        if config.aot_inductor.use_runtime_constant_folding:
```
Also, not sure if this is overkill, but we could add a JK here to fallback for freezing if we run into prod issues

masnesral replied:
Sure I can easily do that if you want. Lemme know if you lean "yes"; I'm leaning against, but I don't have an argument. Side node: TBH, I'm more concerned that the refactoring has the chance of breaking some corner case in caching unrelated to freezing, e.g., the part about how we attach the constant attributes. But it's not practical to killswitch those changes. So I'm running a lot more testing locally, i.e., disabling the "fresh" cache in the inductor tests and running them twice.

oulgen replied:
Can we also run Ads P9 tests (now that it supports warm start on B side)?

The thing that scares me the most is if something breaks with caching, then people will ask us to turn off all of caching instead of rolling back training platform.

Definitely add a JK for freezing

eellison replied:
freezing only applies to inference, and pretty sparsely used internally. not sure we need to run ads training

masnesral replied:
added

masnesral commented:
Doesn't seem to move any metrics (which is expected, so good): https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2012%20Sep%202024%2019%3A58%3A01%20GMT&stopTime=Thu%2C%2026%20Sep%202024%2019%3A58%3A01%20GMT&granularity=hour&suite=torchbench&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/masnesral/116/head&lCommit=8f0769cf701141bce60b36020073f0cc6e218668&rBranch=main&rCommit=db80b98ec460ca5b2fd84c1dfb6426925f64c8cc

masnesral committed:
rebase + merge + comments on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63542170).

masnesral commented:
> Can we also run Ads P9 tests (now that it supports warm start on B side)?

@oulgen I added results from the ads p9 tests on the internal diff. I'm not sure I really trust the cache hit/miss results. On one run, I get more misses on B vs. A. https://fburl.com/servicelab/hk5acdvn. But after a retry, I get the same: https://fburl.com/servicelab/sdb2295s. Modulo that inconsistency, I think the results look fine.

masnesral commented:
@pytorchbot merge

oulgen commented:
Cache hit/miss are not consistent due to different inputs on subsequent attempts :(

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral commented:
@jamesjwu does anything here look problematic for aot autograd caching?

jamesjwu commented on torch/_inductor/codecache.py:
```diff
@@ -807,6 +800,7 @@ def compiled_fx_graph_hash(
 
 
 def cudagraph_post_compile(
+    gm: torch.fx.GraphModule,
```
Just confirming, are we sure the constants on the gm always available post dynamo, and won't change during autograd or inductor compile? As long as they're avaliable on the dynamo generated forward graph for both the forward and the backward, we can use it in AOTAutogradCache.

Are the constants the same for the forward *and* backward graphs? Since we'll be passing the same ones to both? 

Either way, I think instead of passing in the netire gm to cudagraph_post_compile, I would prefer if you passed only the constants in a list as an argument, and moved the get_constants call to be at the caller of the function.

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

jamesjwu commented on torch/_inductor/codecache.py:
```diff
@@ -1361,6 +1365,7 @@ def get_remote_cache() -> Optional[RemoteCache[JsonDataTy]]:
     @staticmethod
     def load_with_key(
         key: str,
+        gm: torch.fx.GraphModule,
```
@masnesral Sorry, the whole point of load_with_key is that you don't pass the `gm` in either haha, do you mind threading it through here too? 

I know the gm is only used for constants here, but it's called in AOTAutogradCache. 

I think in my head, the way it works is taht both AOTAutogradCache.load() and FXGraphCache.load() calculate constants off of their respective gms and then call FXGraphCache.load_with_key()

masnesral replied:
@jamejwu, sorry I guess I'm probably not really following what you're trying to achieve. The primary change in this PR is to stop storing the constants in the cache entry. i.e., we don't have the constants available when we create the module here: https://github.com/pytorch/pytorch/blob/08ce3aac6222bc044a34289be0885e37229d98a2/torch/_inductor/codecache.py#L1124

Instead, when we synthesize the module from the cache, we retrieve the constants from the GraphModule. So I don't understand how we can not pass the GraphModule on the cache lookup path. We can't extract the constants first, because we won't know the attribute _names_ until we have a cache hit and get the names from the cache entry. If the goal is to pass the bare minimum, then I think we'd need to just pass all of the GraphModule's attributes instead of the whole GraphModule object. But that looks a bit strange to me. Lemme know if I totally misunderstand your ask.

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation, consider whether the constant will be inlined or not. If inlined, then we need to include the values in the hash. If not, then we consider only the metadata.
2) Instead of storing the constant attributes in the cache entry, store the _names_ of the constants and then grab those constants from the GraphModule when we need add the constants as attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places, including in autograd_cache code.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along..

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -628,14 +646,80 @@ def f(x, val):
+        with torch.no_grad():
+            x = torch.rand(8, 8).to(device=device, dtype=dtype)
+            out0 = mod1(x)
+            out1 = torch.compile(mod1)(x)
```
@eellison Is there a straightforward way to assert that the freezing opt actually ran? (I verified manually in the debugger, but it would be nice to harden the the test)

eellison replied:
answered offline but - the tests check for `frozen_param`. you could also add a counter bump.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -800,6 +811,10 @@ def _get_custom_pass_detail(
         return custom_pass.uuid()
 
 
+def has_frozen_params(gm: torch.fx.GraphModule) -> bool:
```
Um, yuck. Is there a better way? This is brittle. I can set an attribute on `torch.fx.GraphModule` when freezing is applied?

eellison replied:
Yes - let's find a different solution. if you were to create a module with an attr `_frozen_param` this could fire. Attr set on the module when freezing is applied is fine with me.

masnesral commented:
@jamesjwu / @eellison, here are the changes we discussed. Now we have no changes to the aot autograd cache.

masnesral committed:
Update on "[fx graph cache] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation: If freezing has not been applied, then keep the existing behavior (pickle the metadata and values). If freezing has been applied, then pickle the values if the constant will be inlined; otherwise, consider only the metadata.
2) If freezing has been applied, modify what we store in the cache: Instead of storing the constant attributes in the cache entry, store the _names_ of the constants, and then grab those constants from the GraphModule when we need attache the attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along.

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

laithsakka commented:
for the failing pr_time test you can just update the expected results

masnesral commented:
@laithsakka 
> for the failing pr_time test you can just update the expected results if the change is ok
No way! I'm fixing.

eellison reviewed (approved):
LGTM - @jamesjwu please take a look as I think most of these changes were motivated from your comments

masnesral committed:
Update on "[fx graph cache] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation: If freezing has not been applied, then keep the existing behavior (pickle the metadata and values). If freezing has been applied, then pickle the values if the constant will be inlined; otherwise, consider only the metadata.
2) If freezing has been applied, modify what we store in the cache: Instead of storing the constant attributes in the cache entry, store the _names_ of the constants, and then grab those constants from the GraphModule when we need attache the attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along.

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral committed:
Update on "[fx graph cache] Support freezing with FX graph caching"


Summary: The main changes to support freezing are:
1) When pickling constant tensors as part of the cache key calculation: If freezing has not been applied, then keep the existing behavior (pickle the metadata and values). If freezing has been applied, then pickle the values if the constant will be inlined; otherwise, consider only the metadata.
2) If freezing has been applied, modify what we store in the cache: Instead of storing the constant attributes in the cache entry, store the _names_ of the constants, and then grab those constants from the GraphModule when we need attache the attributes to a newly-loaded Python module. Since the cache lookup path loads the Python module, this bullet means we need to thread through a GraphModule argument in several places.
3) Since this feature means that we may need to reload the same Python module path more than once (but attach different constant attributes), I changed PyCodeCache.load_by_key_path to not store an in-memory map of path to module (since there may be more than one). I don't _think_ this will have any affect on performance, however.. It's unclear why we were using an in-memory cache here anyway, since this function should only be called once for each module needed to be loaded.
4) Several tests were removing on-disk PyCodeCache artifacts by iterating over the modules. I made this more straightforward by implementing a cache_clear method that removes the on-disk artifacts. Arguably, this should have been the implementation all along.

Differential Revision: [D63542170](https://our.internmc.facebook.com/intern/diff/D63542170)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D63542170).

masnesral commented:
@pytorchbot merge
